### PR TITLE
Add currency to create asset profile dialog #3773

### PR DESCRIPTION
**Pull request by madhab-chandra-sahoo**
**[Feature] Integrate add currency into create asset profile dialog #3773**

Screenshot 2024-09-25 at 11 37 18 PM

Screenshot 2024-09-25 at 11 37 43 PM

Screenshot 2024-09-25 at 11 38 11 PM

Screenshot 2024-09-26 at 12 20 16 AM

### CODE REVIEW

File: apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts [hunk 1]
```diff
@@ -51,42 +67,93 @@ export class CreateAssetProfileDialog implements OnInit, OnDestroy {
...
+    } else if (this.mode === 'manual') {
+      this.dialogRef.close({
+        dataSource: 'MANUAL',
+        symbol: this.createAssetProfileForm.get('addSymbol').value
+      });
+    } else {
+      const currency = this.createAssetProfileForm
+        .get('addCurrency')
+        .value.toUpperCase();
+      if (currency.length === 3) {
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts:90]: Add line break

**Review by dtslvr — CHANGES_REQUESTED** [review]
Thank you for your contribution, @madhab-chandra-sahoo! 🙂
I have added a few comments.

**Comment by madhab-chandra-sahoo** [comment]
@dtslvr Thanks for the review. I've updated the code as per your suggestions, Please recheck.

File: apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts [hunk 2]
```diff
@@ -25,17 +32,29 @@ import { MatDialogRef } from '@angular/material/dialog';
...
   public createAssetProfileForm: FormGroup;
-  public mode: 'auto' | 'manual';
+  public mode: 'auto' | 'currency' | 'manual';
+  public customCurrencies: string[];
+
+  private unsubscribeSubject = new Subject<void>();
 
   public constructor(
     public readonly adminService: AdminService,
+    private changeDetectorRef: ChangeDetectorRef,
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts:42]: Please add `readonly`

File: apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.html [hunk 3]
```diff
@@ -37,6 +40,25 @@ <h1 i18n mat-dialog-title>Add Asset Profile</h1>
           <input formControlName="addSymbol" matInput />
         </mat-form-field>
       </div>
+    } @else if (mode === 'currency') {
+      <div>
+        <mat-form-field appearance="outline" class="w-100">
+          <mat-label i18n>Currency</mat-label>
+          <input formControlName="addCurrency" matInput />
+          <mat-error
+            *ngIf="
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.html:49]: Please use the new control flow syntax:

```
@if (createAssetProfileForm.get('addCurrency').hasError('required')) {
  <mat-error i18n>Currency is required.</mat-error>
}
```

File: apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.html [hunk 4]
```diff
@@ -37,6 +40,25 @@ <h1 i18n mat-dialog-title>Add Asset Profile</h1>
...
+              createAssetProfileForm.get('addCurrency').hasError('required')
+            "
+            i18n
+          >
+            Currency is required.
+          </mat-error>
+          <mat-error *ngIf="showCurrencyErrorMessage" i18n>
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.html:56]: Please use the new control flow syntax

File: package.json [hunk 5]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "ghostfolio",
-  "version": "2.110.0",
+  "version": "2.111.0",
```
Comment (dtslvr) [inline package.json:3]: Please keep at `2.110.0`, the version will be increased in the release process 🙂

File: apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts [hunk 6]
```diff
@@ -51,34 +70,56 @@ export class CreateAssetProfileDialog implements OnInit, OnDestroy {
...
+        dataSource:
+          this.createAssetProfileForm.get('searchSymbol').value.dataSource,
+        symbol: this.createAssetProfileForm.get('searchSymbol').value.symbol
+      });
+    } else if (this.mode === 'manual') {
+      this.dialogRef.close({
+        dataSource: 'MANUAL',
+        symbol: this.createAssetProfileForm.get('addSymbol').value
+      });
+    } else {
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts:89]: `} else if (this.mode === 'currency') {`

File: apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts [hunk 7]
```diff
@@ -89,4 +130,39 @@ export class CreateAssetProfileDialog implements OnInit, OnDestroy {
...
+      })
+      .pipe(takeUntil(this.unsubscribeSubject))
+      .subscribe(() => {
+        setTimeout(() => {
+          window.location.reload();
+        }, 300);
+      });
+  }
+
+  get showCurrencyErrorMessage() {
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts:157]:
```
public get showCurrencyErrorMessage() {
```

Please also move up to the other public functions

File: apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts [hunk 8]
```diff
@@ -89,4 +130,39 @@ export class CreateAssetProfileDialog implements OnInit, OnDestroy {
...
+    const addCurrencyFormControl =
+      this.createAssetProfileForm.get('addCurrency');
+    if (
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts:160]: Add line break

File: apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts [hunk 9]
```diff
@@ -89,4 +130,39 @@ export class CreateAssetProfileDialog implements OnInit, OnDestroy {
...
+      addCurrencyFormControl.hasError('minlength') ||
+      addCurrencyFormControl.hasError('maxlength')
+    ) {
+      return true;
+    }
+    return false;
```
Comment (dtslvr) [inline apps/client/src/app/components/admin-market-data/create-asset-profile-dialog/create-asset-profile-dialog.component.ts:166]: Add line break

**Review by dtslvr — CHANGES_REQUESTED** [review]
Thanks a lot for the update! 🤓
I have reviewed again and added a few comments. The issue is almost complete.

**Review by dtslvr — APPROVED** [review]
Thanks again, @madhab-chandra-sahoo!
Ready to merge now 😎

**Pull request merged**